File: apps/api/src/microlab_api/domain/circuit/netlist.py
"""Построение netlist: объединение выводов в электрические узлы (nets).

Узлы образуют:

* соединения (провода) документа;
* внутренние соединения определений (несколько выводов GND платы, полосы макетной платы);
* совпадение по сетке: вывод компонента, который лежит точно в той же точке сетки, что и
  вывод компонента-гнезда (``socket``, например отверстие макетной платы), соединён с
  ним. Выводы платы и выводы других гнёзд так не соединяются. Близость без точного
  совпадения координат соединением не является.

Выводы без соединений в netlist не попадают; узлы, состоящие только из выводов гнёзд
(например, пустая полоса макетной платы), тоже не выводятся.

Порядок детерминирован и не зависит от порядка соединений в документе: вывод
упорядочивается по (позиция объекта в документе — сначала плата, затем компоненты;
позиция вывода в определении), узел — по своему первому выводу. Узлы нумеруются
``NET_001``, ``NET_002``, … в этом порядке.
"""

from dataclasses import dataclass

from microlab_api.circuit_schema.definitions import DefinitionRegistry
from microlab_api.circuit_schema.generated.circuit import CircuitDocument
from microlab_api.domain.circuit.geometry import Point, pin_positions

type PinKey = tuple[int, int]


@dataclass(frozen=True, slots=True)
class Net:
    id: str
    members: tuple[str, ...]
    """Ссылки на выводы вида ``componentId.pinId``, в каноническом порядке."""


class _UnionFind:
    def __init__(self) -> None:
        self._parent: dict[str, str] = {}

    def add(self, item: str) -> None:
        self._parent.setdefault(item, item)

    def find(self, item: str) -> str:
        root = item
        while self._parent[root] != root:
            root = self._parent[root]
        while self._parent[item] != root:
            self._parent[item], item = root, self._parent[item]
        return root

    def union(self, a: str, b: str) -> None:
        self.add(a)
        self.add(b)
        root_a, root_b = self.find(a), self.find(b)
        if root_a != root_b:
            self._parent[root_b] = root_a

    def groups(self) -> list[list[str]]:
        result: dict[str, list[str]] = {}
        for item in self._parent:
            result.setdefault(self.find(item), []).append(item)
        return list(result.values())


def _connect_by_coincidence(
    placements: list[tuple[bool, list[tuple[str, Point]]]], union_find: _UnionFind
) -> set[str]:
    """Соединяет выводы компонентов с выводами гнёзд в той же точке сетки.

    Возвращает множество выводов гнёзд.
    """
    holes: dict[Point, list[str]] = {}
    for is_socket, placed in placements:
        if is_socket:
            for ref, point in placed:
                holes.setdefault(point, []).append(ref)
    for is_socket, placed in placements:
        if not is_socket:
            for ref, point in placed:
                for hole in holes.get(point, ()):
                    union_find.union(ref, hole)
    return {ref for refs in holes.values() for ref in refs}


def build_netlist(document: CircuitDocument, registry: DefinitionRegistry) -> list[Net]:
    """Строит nets документа.

    Ожидает документ без ошибок ``check_references``: соединения с неизвестными
    объектами или выводами пропускаются, повторяющиеся id объектов учитываются по
    первому вхождению.
    """
    board = document.board
    instances = [(board.id, board.type, board.position, board.rotation)]
    instances += [(c.id, c.type, c.position, c.rotation) for c in document.components]

    order: dict[str, PinKey] = {}
    placements: list[tuple[bool, list[tuple[str, Point]]]] = []
    seen: set[str] = set()
    union_find = _UnionFind()
    for index, (instance_id, component_type, position, rotation) in enumerate(instances):
        definition = registry.get(component_type)
        if definition is None or instance_id in seen:
            continue
        seen.add(instance_id)
        for pin_index, pin in enumerate(definition.pins):
            order[f"{instance_id}.{pin.id}"] = (index, pin_index)
        for group in definition.internal_connections or []:
            first, *rest = (f"{instance_id}.{pin_id}" for pin_id in group.root)
            for other in rest:
                union_find.union(first, other)
        if definition.category == "board":
            continue
        placed = [
            (f"{instance_id}.{pin_id}", point)
            for pin_id, point in pin_positions(definition, position, rotation).items()
        ]
        placements.append((bool(definition.socket), placed))

    socket_pins = _connect_by_coincidence(placements, union_find)

    for connection in document.connections:
        a = f"{connection.from_.component_id}.{connection.from_.pin_id}"
        b = f"{connection.to.component_id}.{connection.to.pin_id}"
        if a in order and b in order:
            union_find.union(a, b)

    groups = [
        sorted(group, key=order.__getitem__)
        for group in union_find.groups()
        if len(group) > 1
        and all(ref in order for ref in group)
        and not all(ref in socket_pins for ref in group)
    ]
    groups.sort(key=lambda group: order[group[0]])
    return [
        Net(id=f"NET_{number:03d}", members=tuple(group))
        for number, group in enumerate(groups, start=1)
    ]

File: apps/web/src/features/circuit-model/netlist.ts
import { getComponentDefinition, type ComponentDefinition } from "@microlab/circuit-schema";

import type { NormalizedCircuit } from "./circuit-document";
import { pinGridPosition, toPlacedInstance, type PlacedInstance } from "./geometry";

/**
 * Netlist схемы: объединение выводов в электрические узлы. Правила и порядок совпадают
 * с backend (общие эталонные примеры в circuit-schema проверяют это в обоих языках):
 *
 * - узлы образуют провода, внутренние соединения определений и совпадение по сетке:
 *   вывод компонента, лежащий точно в точке вывода компонента-гнезда (`socket`,
 *   например отверстие макетной платы), соединён с ним; выводы платы и других гнёзд
 *   так не соединяются, близость без совпадения — не соединение;
 * - вывод упорядочивается по (позиция объекта: плата, затем компоненты; позиция вывода
 *   в определении), узел — по первому выводу; номера NET_001, NET_002, …;
 * - выводятся узлы из двух и более выводов, кроме узлов только из выводов гнёзд.
 */

export interface Net {
  id: string;
  /** Ссылки на выводы `componentId.pinId` в каноническом порядке. */
  members: readonly string[];
}

export interface Netlist {
  nets: readonly Net[];
  /** Узел по ключу вывода (только выводы, попавшие в узлы). */
  netByPin: ReadonlyMap<string, Net>;
}

class UnionFind {
  private readonly parent = new Map<string, string>();

  add(item: string): void {
    if (!this.parent.has(item)) this.parent.set(item, item);
  }

  find(item: string): string {
    let root = item;
    for (let next = this.parent.get(root); next !== undefined && next !== root; next = this.parent.get(root)) {
      root = next;
    }
    let current = item;
    while (current !== root) {
      const next = this.parent.get(current) ?? root;
      this.parent.set(current, root);
      current = next;
    }
    return root;
  }

  union(a: string, b: string): void {
    this.add(a);
    this.add(b);
    const rootA = this.find(a);
    const rootB = this.find(b);
    if (rootA !== rootB) this.parent.set(rootB, rootA);
  }

  groups(): string[][] {
    const result = new Map<string, string[]>();
    for (const item of this.parent.keys()) {
      const root = this.find(item);
      const group = result.get(root);
      if (group === undefined) result.set(root, [item]);
      else group.push(item);
    }
    return [...result.values()];
  }
}

type NetlistSource = Pick<
  NormalizedCircuit,
  "board" | "components" | "componentOrder" | "connections" | "connectionOrder"
>;

function pointKey(x: number, y: number): string {
  return `${x},${y}`;
}

export function buildNetlist(circuit: NetlistSource): Netlist {
  const instances: PlacedInstance[] = [toPlacedInstance(circuit.board)];
  for (const id of circuit.componentOrder) {
    const component = circuit.components[id];
    if (component !== undefined) instances.push(component);
  }

  // Порядок вывода: [индекс объекта, индекс вывода].
  const order = new Map<string, readonly [number, number]>();
  const socketPins = new Set<string>();
  const socketHoles = new Map<string, string[]>();
  const plugged: [string, string][] = [];
  const seen = new Set<string>();
  const unionFind = new UnionFind();

  instances.forEach((instance, index) => {
    const definition: ComponentDefinition | undefined = getComponentDefinition(instance.type);
    if (definition === undefined || seen.has(instance.id)) return;
    seen.add(instance.id);
    definition.pins.forEach((pin, pinIndex) => {
      order.set(`${instance.id}.${pin.id}`, [index, pinIndex]);
    });
    for (const group of definition.internalConnections ?? []) {
      const [first, ...rest] = group.map((pinId) => `${instance.id}.${pinId}`);
      if (first === undefined) continue;
      for (const other of rest) unionFind.union(first, other);
    }
    if (definition.category === "board") return;
    for (const pin of definition.pins) {
      const point = pinGridPosition(instance, definition, pin.id);
      if (point === undefined) continue;
      const ref = `${instance.id}.${pin.id}`;
      const key = pointKey(point.x, point.y);
      if (definition.socket === true) {
        socketPins.add(ref);
        const holes = socketHoles.get(key);
        if (holes === undefined) socketHoles.set(key, [ref]);
        else holes.push(ref);
      } else {
        plugged.push([ref, key]);
      }
    }
  });

  for (const [ref, key] of plugged) {
    for (const hole of socketHoles.get(key) ?? []) unionFind.union(ref, hole);
  }

  for (const id of circuit.connectionOrder) {
    const connection = circuit.connections[id];
    if (connection === undefined) continue;
    const a = `${connection.from.componentId}.${connection.from.pinId}`;
    const b = `${connection.to.componentId}.${connection.to.pinId}`;
    if (order.has(a) && order.has(b)) unionFind.union(a, b);
  }

  const compare = (a: string, b: string): number => {
    const [ai, ap] = order.get(a) ?? [0, 0];
    const [bi, bp] = order.get(b) ?? [0, 0];
    return ai - bi || ap - bp;
  };
  const groups = unionFind
    .groups()
    .filter(
      (group) =>
        group.length > 1 && group.every((ref) => order.has(ref)) && !group.every((ref) => socketPins.has(ref)),
    )
    .map((group) => group.sort(compare));
  groups.sort((a, b) => compare(a[0] ?? "", b[0] ?? ""));

  const nets: Net[] = groups.map((members, index) => ({
    id: `NET_${String(index + 1).padStart(3, "0")}`,
    members,
  }));
  const netByPin = new Map<string, Net>();
  for (const net of nets) {
    for (const member of net.members) netByPin.set(member, net);
  }
  return { nets, netByPin };
}

const netlistCache = new WeakMap<object, { circuit: NetlistSource; netlist: Netlist }>();

/**
 * Netlist с кэшем: пересчитывается, только если изменились плата, компоненты или
 * соединения (сравнение по ссылкам нормализованного состояния).
 */
export function getNetlist(circuit: NetlistSource): Netlist {
  const cached = netlistCache.get(circuit.components);
  if (
    cached?.circuit.board === circuit.board &&
    cached.circuit.componentOrder === circuit.componentOrder &&
    cached.circuit.connections === circuit.connections &&
    cached.circuit.connectionOrder === circuit.connectionOrder
  ) {
    return cached.netlist;
  }
  const netlist = buildNetlist(circuit);
  netlistCache.set(circuit.components, {
    circuit: {
      board: circuit.board,
      components: circuit.components,
      componentOrder: circuit.componentOrder,
      connections: circuit.connections,
      connectionOrder: circuit.connectionOrder,
    },
    netlist,
  });
  return netlist;
}
